File: tests/serial.js
'use strict';

var getChiSquared = require('./libs/chi-squared');

function getSerialTest(generator, options) {
	var observations = {};
	var observationType;

    var pairsNumber = Math.floor(options.sequenceLength / 2);

	for (var i = 0; i < pairsNumber; i++) {
		observationType = generator.next() + ';' + generator.next();

		if (observations[observationType] === undefined) {
			observations[observationType] = 0;
		}

		observations[observationType]++;
	}

    var groupsNumber = options.valuesNumber * options.valuesNumber;
    var expectedObservations = pairsNumber / groupsNumber;

	return getChiSquared(observations, expectedObservations, groupsNumber);
}

module.exports = getSerialTest;

File: tests/serial.py
from libs.chi_squared import getChiSquared
import math

def getSerialTest(generator, options):
    observations = {}
    observationType = None

    pairsNumber = int(math.floor(float(options['sequenceLength']) / 2))

    for i in range(pairsNumber):
        observationType = str(generator.next()) + ';' + str(generator.next())

        if not observationType in observations:
            observations[observationType] = 0

        observations[observationType] += 1

    groupsNumber = options['valuesNumber'] * options['valuesNumber']
    expectedObservations = float(pairsNumber) / groupsNumber

    return getChiSquared(observations, expectedObservations, groupsNumber);
